### DISCUSSION TITLE
Docs: Transaction Contexts

### DISCUSSION THREAD

**Discussion by fabiog1901** [psycopg/psycopg#245] [Ideas]
Hello Daniele,

I'm reading the [Transaction Context](https://www.psycopg.org/psycopg3/docs/basic/transactions.html#transaction-contexts) and I love the feature, so neat. I'd like to ask if it is possible to augment the docs with information on whether the below way is correct or not, that is, creating the Cursor obj within the Transaction context. Does it matter, any unexpected behavior, etc..

```python
with psycopg.connect(dburl, autocommit=True) as conn:
    with conn.transaction() as tx:
        with conn.cursor() as cur:
            stmt = """
            insert into orders 
               values (
                        %s, %s, %s, %s, %s,
                        )
                """
            cur.execute(stmt, (
                    self.a, 
                    self.b,
                    self.c, 
                    self.d, 
                    self.e,
                    )
                )
```

or is the below a better approach (transaction context within cursor context) ? 

```python
with psycopg.connect(dburl, autocommit=True) as conn:
    with conn.cursor() as cur:
        with conn.transaction() as tx:
            stmt = """
            insert into orders 
               values (
                        %s, %s, %s, %s, %s,
                        )
                """
            cur.execute(stmt, (
                    self.a, 
                    self.b,
                    self.c, 
                    self.d, 
                    self.e,
                    )
                )
```

Thanks again for looking into this.

Fabio

**Comment by dvarrazzo** [comment]
The two are unrelated. Transactions belong to the connection, cursors operate in the connection whatever its transaction state is. So I would say that "a transaction is bigger than a cursor", conceptually. Closing a cursor (explicitly, on block exit, by gc...) doesn't affect the connection or its transaction state in any way: a (normal) cursor is a purely client-side construct.

My personal pattern, writing a program, is to write functions that take a connection as argument and pass a connection each others, if they have to operate in the same connection/transaction. In this paradigm, some outermost function would deal with the transaction (even at the most basic level of creating and then disposing of the connection), the innermost ones would receive the connection as a parameter and may create their cursors perform their work. Cursors are lightweight so it's no problem to create one to run some command and then discard it. So, in this case too, a transaction block might be outermost and cursors blocks innermost.

Looking at the examples in the docs page, the cursor block is not mentioned at all. The same cursor, in the examples is also used both outside and inside a transaction block. I think it's ok that way, and mentioning the cursor as a block would give the impression that there is a suggested/enforced hierarchical relation between the two objects and I don't think that idea is correct.

So, I would prefer to not change the docs, or, at least, I don't have a personal idea about how to improve them. I am happy to leave this ticket open (I'll convert it into a discussion) and I'm happy to review and merge concrete suggestions about how to improve the docs, if they emerge.
